### PR TITLE
`OnMutate` Observers/`on_mutate` Hooks

### PR DESCRIPTION
**Pull request by grace125**
# Objective

Adopts #14520 and adds `OnMutate` observers/`on_mutate` hooks. 

## Solution

In the `apply` method of each query, for each query with mutable access, runs through the potential components that the query may have mutated, and adds them to a buffer resource. These mutations are flushed in the next command flush.

## Testing

TODO

---

## Showcase

TODO

## TODO

- [ ] Add on_mutate hooks
- [ ] Document code/remove warnings
- [ ] Port over Alice's examples from previous PR; remove temp 
- [ ] Benchmark changes

<details>
<summary>Template</summary>

- Did you test these changes? If so, how?
- Are there any parts that need more testing?
- How can other people (reviewers) test your changes? Is there anything specific they need to know?

## Showcase

> This section is optional. If this PR does not include a visual change or does not add a new feature, you can delete this section.

- Help others understand the result of this PR by showcasing your awesome work!
- If this PR adds a new feature or public API, consider adding a brief pseudo-code snippet of it in action
- If this PR includes a visual change, consider adding a screenshot, GIF, or video
  - If you want, you could even include a before/after comparison!
- If the Migration Guide adequately covers the changes, you can delete this section

While a showcase should aim to be brief and digestible, you can use a toggleable section to save space on longer showcases:

<details>
  <summary>Click to view showcase</summary>

```rust
println!("My super cool code.");
```

</details>
</details>

### CODE REVIEW

**Comment by grace125** [comment]
I'm not sure why "Files Changed" is listing 95 files, sorry if that makes it harder to review. Is it because I squash merged from main?

**Comment by alice-i-cecile** [comment]
> I'm not sure why "Files Changed" is listing 95 files, sorry if that makes it harder to review. Is it because I squash merged from main?

Yeah, it looks like you might have squash-merged in a weird variant of main, because there's merge conflicts now. Don't feel bad if you want to just reopen a fresh PR rather than fighting git.

File: crates/bevy_ecs/src/system/system_param.rs [hunk 1]
```diff
@@ -320,8 +320,94 @@ unsafe impl<D: QueryData + 'static, F: QueryFilter + 'static> SystemParam for Qu
...
+                
+                for archetype_entity in archetype.entities() {
+                    let entity = archetype_entity.id();
+                    let entity_ref = world.entity(entity);
+                    let mutated_observers = mutable_components
+                        .iter()
+                        .cloned()
+                        .filter(|c| entity_ref
+                            .get_change_ticks_by_id(*c)
+                            .map(|ticks| ticks.changed.get() >= system_meta.last_run.get()) // FIXME GRACE: get this_run
```
Comment (grace125) [inline crates/bevy_ecs/src/system/system_param.rs:344]: TODO: don't compare change detection without this_run

File: crates/bevy_ecs/src/system/system_param.rs [hunk 2]
```diff
@@ -320,8 +320,94 @@ unsafe impl<D: QueryData + 'static, F: QueryFilter + 'static> SystemParam for Qu
...
+                            .unwrap_or(false)
+                        ); 
+                    let values = buffer.0.entry(entity).or_default();
+                    values.extend(mutated_observers);
+                }
+            }
+
+            if buffer_started_empty && !buffer.0.is_empty() {
+                world.commands().queue(flush_mutations);
```
Comment (grace125) [inline crates/bevy_ecs/src/system/system_param.rs:353]: This is a jank way that I'm deduplicating `flush_mutations` commands, where I'm only adding a `flush_mutations` command to the queue when `ComponentMutationBuffer` starts empty and has something added to it. Is there a better way of doing this?

**Review by grace125 — COMMENTED** [review]
@alice-i-cecile Some initial comments, some todos and some questions. I probably won't get another chance to work on this again until next weekend, so no rush to respond. Please let me know of any other issues with my solution that you see.

**Comment by grace125** [comment]
Oh also @alice-i-cecile, what's the proper procedure for adopting a PR? Is it fine if I just copy over your examples as long as I cite you in the PR description?

**Comment by alice-i-cecile** [comment]
> Oh also @alice-i-cecile, what's the proper procedure for adopting a PR? Is it fine if I just copy over your examples as long as I cite you in the PR description?

That's fine by me :) You can also use ["co-authored by"](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) in the git commit message to properly credit it in the history, but I personally don't care.

**Comment by alice-i-cecile** [comment]
I'm going to collect the effort / discussion into #16183, which is a bit further along.

**Pull request closed without merging**